src/shogun/base/SGObject.h

```diff
@@ -1129,6 +1129,37 @@ class SGObject: public std::enable_shared_from_this<SGObject>
 		this->observe(obs);
 	}
 
+	/**
+	 * Observe method which emits a function rather than a standard type. The
+	 * emitted function is also wrapped inside another lambda for practicality.
+	 * @tparam T function return type
+	 * @param step time step
+	 * @param name name of this function
+	 * @param method the function we want to emit
+	 */
+	template <class T>
+	void observe(
+		const int64_t step, std::string_view name,
+		std::function<T()> method) const
+	{
+		if (get_num_subscriptions() == 0)
```

Comment:
I guess we could have more involved filters here right?
Like. "check if user requested to observe this name"?

Reply:
mmh it won't be exactly trivial to do so. SGObject does not store directly the observers, just their "subscriptions". The filtering is done automatically by RxCpp (have a look at the `SGObject::subscribe`).

Reply:
I see....then we are kind of back to square one right?
At least the code is now here rather than in the rf model ....

Reply:
Well, the main difference is that we can at least limit the execution of expensive functions now. 

For instance, in the BaggingMachine case, the `compute_oob()` method was called before emitting all observations. Now, the `compute_oob()` is called only if an observable wants it and it is executed by the observable rather than the model.

Reply:
ah maybe I got that wrong then. Cool! :)